### ISSUE TITLE
Properties with type other than "String?"

### ISSUE THREAD

**Issue by NekoiNemo** [jonnyzzz/kotlin.xml.bind#6]
For the life of me i can't figure out how to make a delegate for a property of types other than "String?". Is this possible? If so, how?

**Comment by jonnyzzz** [comment]
No. Currently only String or a sub-object ( via XSub ) are allowed. You are
more than welcome to contribute the support for primitive types.

**Comment by NekoiNemo** [comment]
Ok~

I'm thinking about something like this:
```
///named attribute
open class XAttribute<T : Any>(
        val name: String,
        override val transformLoad: (String?) -> T?,
        override val transformSave: (T) -> String = { it.toString() }
) : Transformable<T> {

  companion object Prebuilt {

    fun String(name: String): XAttribute<String> = XAttribute(name, { it }, { it })
    fun Char(name: String): XAttribute<Char> = XAttribute(name, { it?.first() })

    fun Byte(name: String): XAttribute<Byte> = XAttribute(name, { it?.toByte() })
    fun Int(name: String): XAttribute<Int> = XAttribute(name, { it?.toInt() })
    fun Long(name: String): XAttribute<Long> = XAttribute(name, { it?.toLong() })
    fun Double(name: String): XAttribute<Double> = XAttribute(name, { it?.toDouble() })
    fun Float(name: String): XAttribute<Float> = XAttribute(name, { it?.toFloat() })
  }
}
```
for the declaration, and simply
```
internal class XmlAttributeBind<T : Any>(val attributeName: String, val transformLoad: (String?) -> T?, val transformSave: (T) -> String) : XPropertyImpl<T>(null) {
  override fun loadImpl(scope: Element?): T? = scope?.getAttributeValue(attributeName, null as String?).let(transformLoad)
  override fun saveImpl(value: T, scope: Element) {
    scope.setAttribute(attributeName, value.let(transformSave))
  }
}
```
in the bind.

So in practice it looks like this:
```
@Test fun test_can_read_attribute_with_transform_from_XML() {
    class Data {
      var X : Int? by JXML / XAttribute.Int("a")
    }

    val d : Data = JDOM.load(jdom("aaa") { attribute("a", "42")} , Data::class.java)

    Assert.assertEquals(d.X, 42)
  }
```

What do you think?

**Comment by jonnyzzz** [comment]
It looks we only need an operator to transform Kotlin delegated property
type.

E.g.

var Foo by JXML / XAttribute("food") % Int

Similarly onBefore and onAfter callbacks can be expressed the same way, e.g.

var bar by JXML / XAttribute ("GG") % XCustom(fromfunction, tofunction)

The benefit from this approach is we need not change the rest of the API or
DSL. We only add an extensions to simplify types transformation

It looks like API functions to convert String to other types should provide
means to handle possible convert exception without breaking whole
serialization or deserialization.

**Comment by NekoiNemo** [comment]
To be fair, i can't really figure out how to do it without modifying pretty much everything in the DSL. The declaration chain ends with the multiple concrete XProperty implementaitons (for name, text, etc) and i'm not sure what class or interface should have this % operator function so it would be visible externally and also have access to the required internal data

===
I slightly reworked my previous version to the point where it retains full compatibility with previous version and all alterations are just replacing <String> with <T> and adding <T : Any> to div() in interfaces.
Here's the fork: https://github.com/NekoiNemo/kotlin.xml.bind

How it looks from the outside:
https://github.com/NekoiNemo/kotlin.xml.bind/blob/master/api/src/main/kotlin/org/jonnyzzz/kotlin/xml/bind/XBind.API.kt

**Comment by jonnyzzz** [comment]
More specifically, we may only need to have yet a similar `minus` operator on `XPropery`, e.g.

```
 operator fun <T:Any,Q:Any> XProperty<T>.minus(event : XTransformCallback<T,Q>) : XProperty<Q>
```
where `XTransformCallback` is the same as XCallback, but with an ability to transform T to Q and Q to T respectively. Probably it's ok to change `XCallback` class to be more abstract, but it looks like an API breaking change.

**Comment by jonnyzzz** [comment]
Thanks for your example! It currently looks like you have a combinatoric explosion as you add type conversions to all possible leaf operators (CDATA, Text, Attribute, ...). This will lead to a huge code duplication, sadly. 

We have the base interface for all properties, called `XProperty<T>`. It looks feasible to use the interface to establish all transformation. 

Of course, we need to have the code for each type (in the way you did) to apply it with the `minus` operator to any `XProperty<T>` of any type, e.g. `List<String>`.

How would you cover the case, say `Int` property fails to deserialize? Where should one have a handler with, say default value?

**Comment by NekoiNemo** [comment]
Wouldn't transforming `XProperty<T>` into `XProperty<Q>` require creating new class of XProperty as all current implementations are `XPropertyImpl<String>` ?

**Comment by NekoiNemo** [comment]
Combinatoric explosion can be avoided by introducing a converter class that encapsulates both transformations and just make a set of prefad converters rather than provide implememntations like i did, true.

**Comment by jonnyzzz** [comment]
I see no problem to wrap a given `XProperty<T>` to have an `XProperty<Q>`. Maybe it'd even work without usage of `XPRopertyImpl` (one need to check if reflection stuff is able to detect instances afterward)

**Comment by jonnyzzz** [comment]
There is `XPropertyDelegate` in JDOM implementation that one can use for it. Sadly it's hidden in the JDOM implementation details

**Comment by jonnyzzz** [comment]
The reflection scanner selects only object delegated properties that implement the `XmlBind` interface. This is JDOM specifics again.
